File: src/main/java/org/cimsbioko/utilities/ConfigUtils.java
package org.cimsbioko.utilities;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;

import android.util.Log;
import org.cimsbioko.App;
import org.cimsbioko.R;
import org.cimsbioko.navconfig.NavigatorConfig;
import org.cimsbioko.navconfig.NavigatorModule;
import org.cimsbioko.navconfig.forms.Binding;

import java.util.*;

public class ConfigUtils {

    public static String TAG = ConfigUtils.class.getSimpleName();

    public static String getResourceString(Context context, int id) {
        return context.getString(id);
    }

    public static SharedPreferences getSharedPrefs(Context ctx) {
        return ctx.getSharedPreferences(ctx.getPackageName() + "_preferences", Context.MODE_PRIVATE);
    }

    public static String getPreferenceString(Context context, int key, String defaultValue) {
        return getSharedPrefs(context).getString(getResourceString(context, key), defaultValue);
    }

    public static String getPreferenceString(Context context, String key, String defaultValue) {
        return getSharedPrefs(context).getString(key, defaultValue);
    }

    public static Set<String> getMultiSelectPreference(Context context, String key, Set<String> defaultValues) {
        return getSharedPrefs(context).getStringSet(key, defaultValues);
    }

    public static Boolean getPreferenceBool(Context context, String key, Boolean defaultValue) {
        return getSharedPrefs(context).getBoolean(key, defaultValue);
    }

    public static String getVersion(Context ctx) throws PackageManager.NameNotFoundException {
        return ctx.getPackageManager().getPackageInfo(ctx.getPackageName(), 0).versionName;
    }

    public static Collection<NavigatorModule> getActiveModules(Context ctx) {
        NavigatorConfig cfg = NavigatorConfig.getInstance();
        List<NavigatorModule> actives = new ArrayList<>();
        Set<String> activeModuleNames = getMultiSelectPreference(ctx, ctx.getString(R.string.active_modules_key), cfg.getModuleNames());
        for (NavigatorModule module : cfg.getModules()) {
            if (activeModuleNames.contains(module.getName())) {
                actives.add(module);
            }
        }
        return actives;
    }

    public static void clearActiveModules() {
        Context ctx = App.getApp();
        getSharedPrefs(ctx).edit().remove(ctx.getString(R.string.active_modules_key)).apply();
        Log.i(TAG, "cleared active modules from preferences");
    }

    public static Collection<NavigatorModule> getActiveModuleForBinding(Context ctx, Binding binding) {
        List<NavigatorModule> modules = new ArrayList<>();
        for (NavigatorModule module : getActiveModules(ctx)) {
            if (module.getBindings().containsKey(binding.getName())) {
                modules.add(module);
            }
        }
        return modules;
    }

    public static String getAppName(Context context) {
        return getResourceString(context, R.string.app_name);
    }

    public static String getAppFullName(Context context) {
        String appName = getAppName(context);
        try {
            return String.format("%s %s", appName, getVersion(context));
        } catch (PackageManager.NameNotFoundException e) {
            return appName;
        }
    }
}

File: src/main/kotlin/org/cimsbioko/utilities/ConfigUtils.kt
package org.cimsbioko.utilities

import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager.NameNotFoundException
import android.util.Log
import org.cimsbioko.App
import org.cimsbioko.R
import org.cimsbioko.navconfig.Binding
import org.cimsbioko.navconfig.NavigatorConfig
import org.cimsbioko.navconfig.NavigatorModule
import java.util.*

object ConfigUtils {

    var TAG: String = ConfigUtils::class.java.simpleName

    fun getResourceString(context: Context, id: Int): String {
        return context.getString(id)
    }

    fun getSharedPrefs(ctx: Context): SharedPreferences {
        return ctx.getSharedPreferences(ctx.packageName + "_preferences", Context.MODE_PRIVATE)
    }

    fun getPreferenceString(context: Context, key: Int, defaultValue: String?): String? {
        return getSharedPrefs(context).getString(getResourceString(context, key), defaultValue)
    }

    fun getPreferenceString(context: Context, key: String, defaultValue: String?): String? {
        return getSharedPrefs(context).getString(key, defaultValue)
    }

    fun getMultiSelectPreference(context: Context, key: String, defaultValues: Set<String>): Set<String> {
        return getSharedPrefs(context).getStringSet(key, defaultValues)!!
    }

    fun getPreferenceBool(context: Context, key: String, defaultValue: Boolean): Boolean {
        return getSharedPrefs(context).getBoolean(key, defaultValue)
    }

    @Throws(NameNotFoundException::class)
    fun getVersion(ctx: Context): String {
        return ctx.packageManager.getPackageInfo(ctx.packageName, 0).versionName
    }

    fun getActiveModules(ctx: Context): Collection<NavigatorModule> = NavigatorConfig.instance.run {
        val activeModuleNames = getMultiSelectPreference(ctx, ctx.getString(R.string.active_modules_key), moduleNames)
        modules.filter { it.name in activeModuleNames }
    }

    fun clearActiveModules() {
        val ctx: Context = App.instance
        getSharedPrefs(ctx).edit().remove(ctx.getString(R.string.active_modules_key)).apply()
        Log.i(TAG, "cleared active modules from preferences")
    }

    fun getActiveModuleForBinding(ctx: Context, binding: Binding): Collection<NavigatorModule> {
        val modules: MutableList<NavigatorModule> = ArrayList()
        for (module in getActiveModules(ctx)) {
            if (module.bindings.containsKey(binding.name)) {
                modules.add(module)
            }
        }
        return modules
    }

    private fun getAppName(context: Context): String {
        return getResourceString(context, R.string.app_name)
    }

    fun getAppFullName(context: Context): String {
        val appName = getAppName(context)
        return try {
            String.format("%s %s", appName, getVersion(context))
        } catch (e: NameNotFoundException) {
            appName
        }
    }
}